{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fe63c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Name: Chang Liu\n",
    "For the code challenges, some of them I did it with python, some with Go. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439670a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "1.1.1 pattern count\n",
    "# Insert your PatternCount function here, along with any subroutines you need\n",
    "def pattern_count(text: str, pattern: str) -> int:\n",
    "    count = 0\n",
    "    for i in range(len(text)-len(pattern)+1):\n",
    "        if text[i:i+len(pattern)] == pattern:\n",
    "            count += 1\n",
    "    return count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f0336a",
   "metadata": {},
   "outputs": [],
   "source": [
    "1.1.2 profile-most probable k-mers\n",
    "def profile_most_probable_kmer(text: str, k: int, profile: list[dict[str, float]]) -> str:\n",
    "    #define a initial probability that cannot be worse\n",
    "    bestprob = -1\n",
    "    best =\"\"\n",
    "    # traverse every possible start point (n-k+1) of the k-mer\n",
    "    num = len(text)-k+1\n",
    "    for i in range(num):\n",
    "        prob = 1.0\n",
    "        motif = text[i:i+k]\n",
    " # calculate the product of every column probability\n",
    "        for j in range(k):\n",
    "            prob *= profile[j][motif[j]]\n",
    "        if prob > bestprob:\n",
    "            best = motif\n",
    "            bestprob = prob\n",
    "    return best\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b1b7022",
   "metadata": {},
   "outputs": [],
   "source": [
    "1.1.3 Randomized Motif Finding 1000 times\n",
    "import sys\n",
    "import random\n",
    "\n",
    "def randomized_motif_search(dna: list[str], k: int, t: int) -> list[str]:\n",
    "    \"\"\"Implements the RandomizedMotifSearch algorithm with pseudocounts.\"\"\"\n",
    "    # define the initial score as the worst case so it can updated in the following steps\n",
    "    bestscore = k*t\n",
    "    bestmotif = \" \"\n",
    "    # run the randomized_motif_search function 1000 times, and keep the motifs that has the best score (Smallest)\n",
    "    for i in range(1000):\n",
    "        motifs = randomized_motif_once(dna, k, t)\n",
    "        if score(motifs) < bestscore:\n",
    "            bestmotif = motifs\n",
    "            bestscore = score(motifs)\n",
    "    return bestmotif\n",
    "\n",
    "# runs a single randomized_motif_search\n",
    "def randomized_motif_once(dna: list[str], k: int, t:int) -> list[str]:\n",
    "    initial = randomizedInitial(dna, k, t)\n",
    "    best = initial\n",
    "    # generate a forever loop, only break when the score doesnâ€™t improve.\n",
    "    while True:\n",
    " # generate the profile matrix for the current motifs\n",
    "        profile = Profile(best)\n",
    " # find out the most probable motifs of each seq based on the profile matrix\n",
    "        motif = Motif(dna, k, profile)\n",
    " # update if score improves\n",
    "        if score(motif) < score(best):\n",
    "            best = motif\n",
    "        else:\n",
    "            break\n",
    "        return motif\n",
    "\n",
    "\n",
    "# generate a random k-mer for each seq of the dna list \n",
    "def randomizedInitial(dna: list[str], k: int, t:int) -> list[str]:\n",
    "    initial = []\n",
    "    # for each DNA, generate a random number as the start point of the random k-mer\n",
    "    # the random number is within [0,n-k+1](inclusive) \n",
    "    for i in range(t):\n",
    "        index = random.randint(0, len(dna[i])-k-1)\n",
    "        text = dna[i]\n",
    "        initial.append(text[index:index+k])\n",
    "    return initial\n",
    "\n",
    "\n",
    "# generate the probability of each nucleotide in each column\n",
    "def Profile(motif: list[str]) -> list[dict[str, float]]:\n",
    "    profile = []\n",
    "    # for each column\n",
    "    for j in range(len(motif[0])):\n",
    "        dictionary = {\"A\":1,\"C\":1,\"G\":1,\"T\":1}\n",
    "        ## considering pseudocount\n",
    "        total = 4 + len(motif)\n",
    "        # for each row, count the frequency of each nucleotide\n",
    "        for i in range(len(motif)):\n",
    "            text = motif[i]\n",
    "            dictionary[motif[i][j]] += 1\n",
    "        # calculate the probability\n",
    "        dictionary[\"A\"] = dictionary[\"A\"]/total\n",
    "        dictionary[\"C\"] = dictionary[\"C\"]/total\n",
    "        dictionary[\"G\"] = dictionary[\"G\"]/total\n",
    "        dictionary[\"T\"] = dictionary[\"T\"]/total\n",
    "        profile.append(dictionary)\n",
    "    return profile\n",
    "\n",
    "\n",
    "# input the profile matrix of a set of motifs and a DNA seq, output the most-probable k-mer\n",
    "def Profile_MostProb_Kmer(text:str,k: int,profile:list[dict[str,float]])-> str:\n",
    "    # define a initial probability that is worse so that it definitely will update in the next step\n",
    "    bestprob = -1\n",
    "    num = len(text)- len(profile) +1\n",
    "    best =\"\"\n",
    "    # traverse every possible start point of the k-mer\n",
    "    for i in range(num):\n",
    "        prob = 1.0\n",
    "        motif = text[i:i+len(profile)]\n",
    "        for j in range(k):\n",
    "    #calculate the product of probability of each column\n",
    "            prob *= profile[j][motif[j]]\n",
    " # output the motifs with the largest probability\n",
    "        if prob > bestprob:\n",
    "            best = motif\n",
    "            bestprob = prob\n",
    "    return best\n",
    "\n",
    "\n",
    "# find out the most-probable motifs for multiple sequences\n",
    "def Motif(dna:list[str], k: int, profile: list[dict[str,float]])->list[str]:\n",
    "    best = []\n",
    "    for i in range(len(dna)):\n",
    "        motif = Profile_MostProb_Kmer(dna[i],k,profile)\n",
    "        best.append(motif)\n",
    "    return best\n",
    "\n",
    "\n",
    "# sum the number of elements that are disagree with the consensus motifs\n",
    "def score(motif:list[str]) -> int:\n",
    "    M_score = 0\n",
    "    # for every column\n",
    "    for j in range(len(motif[0])):\n",
    "        dictionary = {\"A\":0,\"C\":0,\"G\":0,\"T\":0}\n",
    "    # for every row, count the frequency of each nucleotide\n",
    "        for i in range(len(motif)):\n",
    "            text = motif[i]\n",
    "            dictionary[text[j]] += 1\n",
    " # the highest frequency is the element of the consensus motifs\n",
    "        consensus = max(dictionary[i] for i in [\"A\",\"C\",\"G\",\"T\"])\n",
    "        different = sum(dictionary[i] for i in [\"A\",\"C\",\"G\",\"T\"]) - consensus\n",
    "        M_score += different\n",
    "    return M_score"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
