{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. find the any substring in a genome context decoding a target amino acid sequence, including gene substring and its reverse complement\n",
    "file_path = \"./dataset_35902_6.txt\"\n",
    "\n",
    "# Open the file in read mode\n",
    "with open(file_path, 'r') as file:\n",
    "    # Read the entire content of the file as a single string\n",
    "    lines = file.readlines()\n",
    "\n",
    "Geneseq = \"\"\n",
    "for i in range(len(lines)-1):\n",
    "    Geneseq += lines[i]\n",
    "aa = lines[-1][:-1]\n",
    "\n",
    "genetic_code = {\n",
    "    'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "    'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "    'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "    'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'UAU': 'Y', 'UAC': 'Y', 'UAA': '', 'UAG': '',\n",
    "    'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'UGU': 'C', 'UGC': 'C', 'UGA': '', 'UGG': 'W',\n",
    "    'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "# write a function to generate all possible combinations of codons\n",
    "def generate_all_combinations(dataDict, seq):\n",
    "    possibleSubstring = []\n",
    "    # base case: there is only the last aa in the seq\n",
    "    if len(seq) == 1:\n",
    "        return [code for code in dataDict[seq]]\n",
    "    firstLetter = seq[0]\n",
    "    remaining_seq = seq[1:]\n",
    "    \n",
    "    remaining_combination = generate_all_combinations(dataDict, remaining_seq)\n",
    "    \n",
    "    for codon in dataDict[firstLetter]:\n",
    "        for combinations in remaining_combination:\n",
    "            substring = codon + combinations\n",
    "            possibleSubstring.append(substring)\n",
    "\n",
    "    return possibleSubstring\n",
    "\n",
    "\n",
    "# a function that transform a given string to its reverse_complement\n",
    "def reverse_complement(stringList):\n",
    "    result = []\n",
    "    baseDict = {\"A\":\"T\",\"T\":\"A\",\"C\":\"G\",\"G\":\"C\"}\n",
    "    for s in stringList:\n",
    "        reverse = s[::-1]\n",
    "        complement = \"\"\n",
    "        for letter in reverse:\n",
    "            letter = baseDict[letter]\n",
    "            complement += letter\n",
    "        result.append(complement)\n",
    "    return result\n",
    "\n",
    "# generate a dictionary store the corresponding codon for each amino acid in the aa seq.\n",
    "codeDict = {}\n",
    "for i in range(len(aa)):\n",
    "    if aa[i] not in codeDict.keys():\n",
    "        # transform from U to T\n",
    "        code = [code.replace(\"U\",\"T\") for code in genetic_code.keys() if genetic_code[code] == aa[i]]\n",
    "        codeDict[aa[i]] = []\n",
    "        codeDict[aa[i]].extend(code)\n",
    "\n",
    "# this is the set of genes that decoding the target aa directly\n",
    "possibleSubstring = generate_all_combinations(codeDict,aa)\n",
    "reverseComplement = reverse_complement(possibleSubstring)\n",
    "possibleSubstring.extend(reverseComplement)\n",
    "\n",
    "\n",
    "substrings = []\n",
    "stringLength = len(aa)*3\n",
    "for i in range(len(Geneseq)-stringLength+1):\n",
    "    if Geneseq[i:i+stringLength] in possibleSubstring:\n",
    "        substring = Geneseq[i:i+stringLength]\n",
    "        substrings.append(substring)\n",
    "\n",
    "for i in substrings:\n",
    "    print(i)\n",
    "\n",
    "# 2. given a cyclic DNA of length n, how many substrings might it generate\n",
    "# it have n start point for substrings of any length, it has n-1 possible lengths for substrings, so the number should be n*(n-1)\n",
    "\n",
    "\n",
    "# 3. generate the cyclospectrum of a given peptide\n",
    "amino_acid_weights = {\n",
    "    'G': 57,\n",
    "    'A': 71,\n",
    "    'S': 87,\n",
    "    'P': 97,\n",
    "    'V': 99,\n",
    "    'T': 101,\n",
    "    'C': 103,\n",
    "    'I': 113,\n",
    "    'L': 113,  # Note: I and L have the same weight\n",
    "    'N': 114,\n",
    "    'D': 115,\n",
    "    'K': 128,\n",
    "    'Q': 128,  # Note: K and Q have the same weight\n",
    "    'E': 129,\n",
    "    'M': 131,\n",
    "    'H': 137,\n",
    "    'F': 147,\n",
    "    'R': 156,\n",
    "    'Y': 163,\n",
    "    'W': 186  \n",
    "}\n",
    "\n",
    "def subPeptides(peptide):\n",
    "    length = len(peptide)\n",
    "    subpep = []\n",
    "    # every start point\n",
    "    for j in range(1, length):\n",
    "        # every possible length of subpeptides\n",
    "        for i in range(length): \n",
    "            if i+j <= length: \n",
    "                subpep.append(peptide[i:i+j])\n",
    "            else:\n",
    "                end = i+j-length\n",
    "                sub = peptide[i:]+peptide[:end]\n",
    "                subpep.append(sub)\n",
    "    return subpep\n",
    "\n",
    "# this is to generate spectrum, all mass of subpep and sort them\n",
    "def massCal(subpeptides, amino_acid_weights):\n",
    "    massList= [0]\n",
    "    for s in subpeptides:\n",
    "        mass = 0\n",
    "        for letter in s:\n",
    "            mass += amino_acid_weights[letter]\n",
    "        massList.append(mass)\n",
    "    return sorted(massList)\n",
    "\n",
    "# file_path = \"./dataset_35904_4.txt\"\n",
    "\n",
    "# # Open the file in read mode\n",
    "# with open(file_path, 'r') as file:\n",
    "#     # Read the entire content of the file as a single string\n",
    "#     peptide = file.read()\n",
    "\n",
    "# peptide = peptide.strip('\\n')\n",
    "# subpeptide = subPeptides(peptide)\n",
    "\n",
    "# subMass = massCal(subpeptide, amino_acid_weights)\n",
    "# mass = 0\n",
    "# for letter in peptide:\n",
    "#     mass += amino_acid_weights[letter]\n",
    "# subMass.append(mass)\n",
    "# result = ' '.join(map(str, subMass))\n",
    "# print(result)\n",
    "\n",
    "\n",
    "#4. for linear subpeptide, it should be the sum from 1 to length(1+2+...+length) +1\n",
    "#5. branching and bounding algo for cyclopeptide sequencing\n",
    "def CyclopeptideSeqencing(Spectrum, amino_acid_weights):\n",
    "    candidatePeptide = {''}\n",
    "    finalPeptide = []\n",
    "\n",
    "    while candidatePeptide:\n",
    "        \n",
    "        # the expand should stop when the peptide has been the same size as the expected one\n",
    "        candidatePeptide = Expand(candidatePeptide.copy(), amino_acid_weights)\n",
    "        pep_to_remove = set()\n",
    "        # if \"\" in candidatePeptide:\n",
    "        #     candidatePeptide.remove(\"\")\n",
    "        for pep in candidatePeptide:\n",
    "            if pep == '':\n",
    "                pep_to_remove.add(pep)\n",
    "                continue\n",
    "\n",
    "            if Mass(pep, amino_acid_weights) == max(Spectrum):\n",
    "                if Cyclospectrum(pep, amino_acid_weights) == Spectrum and pep not in finalPeptide:\n",
    "                    finalPeptide.append(pep)\n",
    "                pep_to_remove.add(pep)\n",
    "            elif not Consistent(Linearspectrum(pep, amino_acid_weights), Spectrum):\n",
    "                pep_to_remove.add(pep)\n",
    "            \n",
    "        candidatePeptide -= pep_to_remove\n",
    "    \n",
    "    return finalPeptide\n",
    "\n",
    "\n",
    "def Mass(pep, amino_acid_weights):\n",
    "    mass = 0\n",
    "    for letter in pep:\n",
    "        mass += amino_acid_weights[letter]\n",
    "    return mass\n",
    "\n",
    "def Cyclospectrum(pep, amino_acid_weights):\n",
    "    sub = subPeptides(pep)\n",
    "    submass = massCal(sub, amino_acid_weights)\n",
    "    submass.append(Mass(pep, amino_acid_weights))\n",
    "    # result = ' '.join(map(str, subMass))\n",
    "\n",
    "    return submass \n",
    "    # return result\n",
    "\n",
    "def Linearspectrum(pep, amino_acid_weights):\n",
    "    \n",
    "    length = len(pep)\n",
    "    subpep = []\n",
    "    # every start point\n",
    "    for j in range(1, length):\n",
    "        # every possible length of subpeptides\n",
    "        for i in range(length): \n",
    "            if i+j <= length: \n",
    "                subpep.append(pep[i:i+j])\n",
    "            \n",
    "    submass = massCal(subpep, amino_acid_weights)\n",
    "    submass.append(Mass(pep, amino_acid_weights))\n",
    "\n",
    "    return submass \n",
    "\n",
    "# massList is a tuple\n",
    "def numSpectrum(massList):\n",
    "    length = len(massList)\n",
    "    spectrumList = [0]\n",
    "    for j in range(1, length+1):\n",
    "        for i in range(length):\n",
    "            if i+j <= length:\n",
    "                subpep = massList[i:i+j]\n",
    "                spectrumList.append(sum(subpep))\n",
    "    return sorted(spectrumList)\n",
    "\n",
    "def Expand(candidate, amino_acid_weights):\n",
    "    newCandidate = set()\n",
    "    for p in candidate:\n",
    "        \n",
    "        for aa in amino_acid_weights.keys():\n",
    "            expanded = p + aa\n",
    "            newCandidate.add(expanded)\n",
    "            \n",
    "    return newCandidate\n",
    "\n",
    "def NumExpand(candidate, new_alphabet):\n",
    "    newCandidate = set()\n",
    "   \n",
    "   # candidate is start from {tuple([])}\n",
    "\n",
    "    for p in candidate:\n",
    "        expanded = []\n",
    "        for i in p:\n",
    "            expanded.append(i)\n",
    "\n",
    "        for c in new_alphabet:\n",
    "            copy = expanded.copy()\n",
    "            copy.append(c)\n",
    "            \n",
    "\n",
    "            newCandidate.add(tuple(copy))\n",
    "    return newCandidate\n",
    "\n",
    "\n",
    "def Consistent(testSpectrum, experiment):\n",
    "    if  set(testSpectrum).issubset(set(experiment)) == False:\n",
    "        return False\n",
    "    else:\n",
    "        countA = {}\n",
    "        for m in testSpectrum:\n",
    "            if m not in countA.keys():\n",
    "                countA[m] = testSpectrum.count(m)\n",
    "        countB = {}\n",
    "        for n in experiment:\n",
    "            if n not in countB.keys():\n",
    "                countB[n] = experiment.count(n)\n",
    "        for keys in countA.keys():\n",
    "            if countA[keys] > countB[keys]:\n",
    "                return False\n",
    "    return True\n",
    "\n",
    "\n",
    "# file_path = 'dataset_35906_6.txt'\n",
    "# with open(file_path,\"r\") as file:\n",
    "#     spectrum = file.read()\n",
    "# spectrum = spectrum.strip('\\n')\n",
    "# spectrum = list(map(int, spectrum.split()))\n",
    "\n",
    "# print(spectrum)\n",
    "# # Join the list elements using a comma delimiter (',')\n",
    "# # result = ','.join(numbers_list)\n",
    "# # spectrum = [0, 113, 128, 186, 241, 299, 314, 427]\n",
    "# subPep = CyclopeptideSeqencing(spectrum, amino_acid_weights)\n",
    "# print(subPep)\n",
    "# # transform to number\n",
    "# massList = set()\n",
    "# for element in subPep:\n",
    "#     expression = ''\n",
    "#     for c in element:\n",
    "#         expression += str(amino_acid_weights[c])\n",
    "#         expression += '-'\n",
    "#     expression = expression[:-1]\n",
    "#     massList.add(expression)\n",
    "\n",
    "# for e in massList:\n",
    "#     print(e)\n",
    "\n",
    "#6. score(peptide, spectrum)\n",
    "# file_path = 'dataset_35907_3.txt'\n",
    "# with open(file_path,\"r\") as file:\n",
    "#     lines = file.readlines()\n",
    "\n",
    "# peptide = lines[0].strip('\\n')\n",
    "# spectrum = []\n",
    "# for i in range(1,len(lines)):\n",
    "#     for number in lines[i].split():\n",
    "#         spectrum.append(int(number))\n",
    "\n",
    "# theoretical = Cyclospectrum(peptide,amino_acid_weights)\n",
    "\n",
    "# # print(spectrum)\n",
    "# score = 0\n",
    "# for n in spectrum.copy():\n",
    "#     if n in theoretical:\n",
    "#         score += 1\n",
    "#         theoretical.remove(n)\n",
    "#         spectrum.remove(n)\n",
    "        \n",
    "# print(score)\n",
    "\n",
    "# linear score\n",
    "def Score(theoretical, spectrum):\n",
    "    # theoretical = Linearspectrum(peptide,amino_acid_weights)\n",
    "    score = 0\n",
    "    Refercopy = spectrum.copy()\n",
    "    testcopy = theoretical.copy()\n",
    "    for n in spectrum:\n",
    "        if n in testcopy:\n",
    "            score += 1\n",
    "            testcopy.remove(n)\n",
    "            Refercopy.remove(n)\n",
    "            \n",
    "    return score\n",
    "\n",
    "# 7. LeaderboardCyclopeptideSequencing(spectrum, N)\n",
    "def LeaderBoardSequencing(spectrum, n):\n",
    "    if not spectrum:  # Check if spectrum is empty\n",
    "        return None \n",
    "    leaderBoard = {''}\n",
    "    leaderPep = ''\n",
    "    pep_to_remove = set()\n",
    "    while leaderBoard:\n",
    "        leaderBoard = Expand(leaderBoard,amino_acid_weights)      \n",
    "        for peptide in leaderBoard:\n",
    "            if Mass(peptide, amino_acid_weights) == max(spectrum):\n",
    "                theoretical = Linearspectrum(peptide,amino_acid_weights)\n",
    "                if Score(theoretical, spectrum) > Score(Linearspectrum(leaderPep,amino_acid_weights), spectrum):\n",
    "                    leaderPep = peptide\n",
    "            elif Mass(peptide, amino_acid_weights) > max(spectrum):\n",
    "                pep_to_remove.add(peptide)\n",
    "        leaderBoard -= pep_to_remove\n",
    "        if len(leaderBoard) > n:    \n",
    "            leaderBoard = Trim(leaderBoard, spectrum, n)\n",
    "    return leaderPep\n",
    "\n",
    "def Trim(board, spectrum, n):\n",
    "    boardList = []\n",
    "    scoreList = []\n",
    "    for pep in board:\n",
    "        boardList.append(pep)\n",
    "        theoretical = Linearspectrum(pep,amino_acid_weights)\n",
    "        scoreList.append(Score(theoretical, spectrum))\n",
    "        \n",
    "    sorted_scores = sorted(scoreList, reverse=True)\n",
    "    # the score at N-th place\n",
    "    # find if there is any tie\n",
    "    Nindex = [x for x in range(len(sorted_scores)) if scoreList[x]>= sorted_scores[n]]\n",
    "    newBoard = []\n",
    "    for x in Nindex:\n",
    "        newBoard.append(boardList[x])\n",
    "    newBoard = set(newBoard)\n",
    "    return newBoard\n",
    "\n",
    "# file_path = 'dataset_35907_8.txt'\n",
    "# with open(file_path,\"r\") as file:\n",
    "#     content = file.readlines()\n",
    "# n = content[0].strip('\\n')\n",
    "# n = int(n)\n",
    "# spectrum = []\n",
    "# for i in range(1,len(content)):\n",
    "#     for number in content[i].split():\n",
    "#         spectrum.append(int(number))\n",
    "# print(n)\n",
    "# result = LeaderBoardSequencing(spectrum, n)\n",
    "# # transform to number\n",
    "# expression = ''\n",
    "# for c in result:\n",
    "#     expression += str(amino_acid_weights[c])\n",
    "#     expression += '-'\n",
    "    \n",
    "# print(expression[:-1])\n",
    "\n",
    "# 8. calculate the convolution of a given spectrum\n",
    "def Convolution(spectrum):\n",
    "    convolution = []\n",
    "    for i in range(len(spectrum)):\n",
    "        for j in range(i+1, len(spectrum)):\n",
    "            differ = spectrum[j]-spectrum[i]\n",
    "            if differ > 0:\n",
    "                convolution.append(differ)\n",
    "    return convolution\n",
    "\n",
    "\n",
    "# file_path = \"./dataset_35909_4.txt\"\n",
    "# with open(file_path, \"r\") as file:\n",
    "#     spectrum = file.read()\n",
    "# spectrum = spectrum.strip('\\n')\n",
    "# spectrum = list(map(int, spectrum.split()))\n",
    "# result = Convolution(spectrum)\n",
    "# output_string = ' '.join(map(str, result))\n",
    "\n",
    "# print(output_string)\n",
    "\n",
    "# 9. ConvolutionCyclopeptideSequencing\n",
    "def ConvolutionCyclopeptideSequencing(m, n, spectrum):\n",
    "    # generate the new alphabet(only include the convolution value that is within 57-200)\n",
    "    convolution = Convolution(spectrum)\n",
    "    valid_convolution = [x for x in convolution if x <= 200 and x >= 57]\n",
    "    count = {key: valid_convolution.count(key) for key in valid_convolution}\n",
    "    cutVal = sorted(count.values(),reverse=True)[m]\n",
    "    new_alphabet = [c for c in count.keys() if count[c] >= cutVal]\n",
    "    \n",
    "    \n",
    "    # below is the leaderboard sequencing part, it is modified due to the new alphabet\n",
    "    leaderBoard = {tuple([])}\n",
    "    \n",
    "    leaderPep = []\n",
    "    pep_to_remove = set()\n",
    "    while leaderBoard:\n",
    "        # leaderBoard is now a set of tuples, rather than a set of strings\n",
    "        leaderBoard = NumExpand(leaderBoard,new_alphabet) \n",
    "        for peptide in leaderBoard:\n",
    "            if sum(peptide) == max(spectrum):\n",
    "                theoretical = numSpectrum(peptide)\n",
    "                if Score(theoretical, spectrum) > Score(numSpectrum(leaderPep), spectrum):\n",
    "                    leaderPep = peptide\n",
    "            elif sum(peptide) > max(spectrum):\n",
    "                pep_to_remove.add(peptide)\n",
    "        leaderBoard -= pep_to_remove\n",
    "        if len(leaderBoard) > n:    \n",
    "            leaderBoard = numTrim(leaderBoard, spectrum, n)\n",
    "    theoretical1 = numSpectrum(list(leaderPep))\n",
    "    print(Score(theoretical1,spectrum))\n",
    "    print(theoretical1)\n",
    "    print(leaderPep)\n",
    "    theoretical2 = numSpectrum([99,71,137,57,72,57])\n",
    "    print(Score(theoretical2,spectrum))\n",
    "    print(theoretical2)\n",
    "    return leaderPep\n",
    "\n",
    "# an updated version of trim function considering board is now a set of tuples\n",
    "def numTrim(board, spectrum, n):\n",
    "    boardList = []\n",
    "    scoreList = []\n",
    "    for pep in board:\n",
    "        boardList.append(pep)\n",
    "        theoretical = numSpectrum(pep)\n",
    "        scoreList.append(Score(theoretical, spectrum))\n",
    "        \n",
    "    sorted_scores = sorted(scoreList, reverse=True)\n",
    "    # the score at N-th place\n",
    "    # find if there is any tie\n",
    "    Nindex = [x for x in range(len(sorted_scores)) if scoreList[x]>= sorted_scores[n]]\n",
    "    newBoard = set()\n",
    "    for x in Nindex:\n",
    "        newBoard.add(boardList[x])\n",
    "    newBoard = set(newBoard)\n",
    "    return newBoard\n",
    "\n",
    "\n",
    "file_path = './dataset_35909_7.txt'\n",
    "with open(file_path,'r') as file:\n",
    "    content = file.readlines()\n",
    "\n",
    "m = content[0]\n",
    "m = int(m)\n",
    "n = content[1]\n",
    "n = int(n)\n",
    "spectrum = []\n",
    "for i in range(2,len(content)):\n",
    "    for number in content[i].split():\n",
    "        spectrum.append(int(number))\n",
    "result = ConvolutionCyclopeptideSequencing(m, n, spectrum)\n",
    "output_string = '-'.join(map(str, result))\n",
    "print(output_string)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
